index.bs

```diff
@@ -1693,6 +1766,114 @@ The [=remote end steps=] with |command parameters| are:
 
 </div>
 
+#### The browsingContext.navigate Command ####  {#command-browsingContext-navigate}
+
+The <dfn export for=commands>browsingContext.navigates</dfn> command navigates a
+browsing context to the given URL.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+      <pre class="cddl remote-cddl">
+      BrowsingContextGetTreeCommand = {
+        method: "browsingContext.navigate",
+        params: BrowsingContextNavigateParameters
+      }
+
+      BrowsingContextNavigateParameters = {
+        context: BrowsingContext,
+        url: text,
+        ?wait: ReadinessState,
+      }
+
+       ReadinessState = "none" / "interactive" / "complete"
+      </pre>
+   </dd>
+   <dt>Return Type</dt>
+   <dd>
+    <pre class="cddl local-cddl">
+        BrowsingContextNavigateResult = {
+            navigation: Navigation / null,
+            url: text,
+        }
+    </pre>
+   </dd>
+</dl>
+
+<div algorithm="remote end steps for browsingContext.navigate">
+The [=remote end steps=] with |command parameters| are:
+
+  1. Let |context id| be the value of the <code>context</code> field of
+     |command parameters|.
+
+  1. Let |context| be the result of [=trying=] to [=get a browsing context=]
+     with |context id|.
+
+  1. Assert: |context| is not null.
+
+  1. Let |wait condition| be the value of the <code>wait</code> field of |command
+     parameters| if present, or "<code>none</code>" otherwise.
+
+  1. Let |url| be the value of the <code>url</code> field of |command
+     parameters|.
+
+  1. TODO: validate |url|. Should we allow relative URLs or only absolute URLs?
+
+  1. Let |request| be a new [=/request=] whose URL is |url|.
+
+  1. Let |navigate status| be the result of [=navigate=] |context| with resource
+     |request|, and using |context| as the [=source browsing context=].
+
+  1. If |navigate status|'s status is complete and |navigate status|'s
+     navigation id is null:
+
+    1. Let |body| be a [=map=] matching the
+       <code>BrowsingContextNavigationResult</code> production, with the
+       <code>navigation</code> field set to null, and the <code>url</code> field
+       set to the result of the [=URL serializer=] given |navigate status|'s url.
+
+    1. Return [=success=] with data |body|
+
+     Note: this is the case if the navigation only caused the fragment to change.
+
+  1. If |navigate status|'s status is cancelled return [=error=] with [=error code=]
+     [=unknown error=].
+
+     TODO: is this the right way to handle errors here?
+
+  1. Assert: |navigate status|'s status is pending and |navigate status|'s id is
+     not null.
+
+  1. Let |navigation id| be |navigate status|'s id
+
+  1. If |wait condition| is "<code>none</code>", return [=success=] with data
+     |body|.
+
+  1. If |wait condition| is interactive, let |event name| be
+     "<code>DOMContentLoaded</code>", otherwise let |event name| be "<code>load</code>".
+
+  1. Let (|event received|, |status|) be [=await=] [(|event name|, |navigation
```

Comment:
This places 3 pairs in the wait queue, but at most 1 of them will be resumed which leaves 2 in the queue indefinitely. Should we define an "await any" primitive in this spec similar in purpose to Promise.race() that lets us express this step more cleanly? Might come in handy as we add more functionality to the spec.

Reply:
For now I'bve changed things so there's a single id for all the cases, so we can make the spec structure a map from id to events. I don't know if that will always work, or if there will be cases where we might get a different id for different events we await (also, it doesn't really matter if we happen to leak in the spec; obviously it's a problem in a real implementation, but hopefully it's not too difficult to infer that the intent here is "wait for whichever of these things happens first" and it's obvious that should translate into `Promise.race` or implementation equivalent.